IdentityCore/src/cache/crypto/MSIDAssymetricKeyKeychainGenerator.m

```diff
@@ -0,0 +1,276 @@
+        return nil;
+    }
+    
+    NSDictionary *privateKeyAttributes = [self keyAttributesWithQueryDictionary:[attributes privateKeyAttributes] keyTitle:@"private key" error:error];
+    
+    if (!privateKeyAttributes)
+    {
+        return nil;
+    }
+    
+    NSDictionary *publicKeyAttributes = [self keyAttributesWithQueryDictionary:[attributes publicKeyAttributes] keyTitle:@"public key" error:error];
+    
+    if (!publicKeyAttributes)
+    {
+        return nil;
+    }
+    
+    SecKeyRef privateKeyRef = (__bridge SecKeyRef)privateKeyAttributes[(__bridge id)kSecValueRef];
+    SecKeyRef publicKeyRef = (__bridge SecKeyRef)publicKeyAttributes[(__bridge id)kSecValueRef];
+    
+    if (!privateKeyRef || !publicKeyRef)
+    {
+        [self logAndFillError:@"Invalid keychain attributes. No key ref returned" status:-1 error:error];
+        return nil;
+    }
+    
+    MSIDAssymetricKeyPair *keypair = [[MSIDAssymetricKeyPair alloc] initWithPrivateKey:privateKeyRef
+                                                                             publicKey:publicKeyRef];
+
+    
+    return keypair;
+}
+
+#pragma mark - Cleanup
+
+- (BOOL)cleanupKeychainForAttributes:(MSIDAssymetricKeyLookupAttributes *)attributes error:(NSError **)error
+{
+    return [self deleteItemWithAttributes:[attributes privateKeyAttributes] itemTitle:@"private key" error:error]
+        && [self deleteItemWithAttributes:[attributes publicKeyAttributes] itemTitle:@"public key" error:error];
+}
+
+- (BOOL)deleteItemWithAttributes:(NSDictionary *)attributes itemTitle:(NSString *)itemTitle error:(NSError **)error
+{
+    NSDictionary *queryAttributes = [self keychainQueryWithAttributes:attributes];
+    OSStatus result = SecItemDelete((CFDictionaryRef)queryAttributes);
+    
+    if (result != errSecSuccess
+        && result != errSecItemNotFound)
+    {
+        [self logAndFillError:[NSString stringWithFormat:@"Failed to remove %@", itemTitle]
+                       status:result
+                        error:error];
+        return NO;
+    }
+    
+    return YES;
+}
+
+#pragma mark - Private
+
+- (NSDictionary *)keyAttributesWithQueryDictionary:(NSDictionary *)queryDictionary keyTitle:(NSString *)keyTitle error:(NSError **)error
+{
+    NSMutableDictionary *keychainQuery = [[self keychainQueryWithAttributes:queryDictionary] mutableCopy];
+    keychainQuery[(__bridge id)kSecReturnAttributes] = @YES;
+    keychainQuery[(__bridge id)kSecReturnRef] = @YES;
+    
+    CFTypeRef keyCFDict = NULL;
+    
+    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)keychainQuery, (CFTypeRef*)&keyCFDict);
+    
+    if (status != errSecSuccess)
+    {
+        if (status != errSecItemNotFound)
+        {
+            [self logAndFillError:[NSString stringWithFormat:@"Failed to find %@", keyTitle]
+                           status:status
+                            error:error];
+        }
+        
+        return nil;
```

Comment:
I think it will be good to just log here as well before returning nil without filling the error for any other keychain errors we might encounter to investigate issues in the future.

Reply:
Sure, will fix :)

Reply:
Addressed.